### ISSUE TITLE
vsnprintf delcaration issue on Big Sur?

### ISSUE THREAD

**Issue by mikeymastro** [sm64pc/sm64pc#462]
Hey everyone.  Bear with me as this stuff is not my forte.  I just came across some tutorials to install sm64 on a MAC and i'm running into the below issue, which lead me here.  I'm kind of stuck.  I read other issues and tried some suggestions, but nothing appears to work.  Any help would be appreciated.  Cheers!

```
gcc -Llib -Iinclude -I../include -I . -Wall -Wextra -Wno-unused-parameter -DOSX_BUILD -pedantic -std=c99 -O2 -s -DOSX_BUILD aiff_extract_codebook.c -lm -o aiff_extract_codebook 
aiff_extract_codebook.c:48:16: error: implicitly declaring library function 'vsnprintf' with type 'int (char *, unsigned long, const char *, struct __va_list_tag *)'
      [-Werror,-Wimplicit-function-declaration]
    int size = vsnprintf(NULL, 0, fmt, ap);
               ^
aiff_extract_codebook.c:48:16: note: include the header <stdio.h> or explicitly provide a declaration for 'vsnprintf'
1 error generated.
```

**Comment by mikeymastro** [comment]
I moved the <stdio.h> to the very top of the c file, even though it was in there, and i ran the compiler again, this time I'm getting the following error:

```
cpp-9 -P -DVERSION_US -DNON_MATCHING -DAVOID_UB -DOSX_BUILD -I . -o build/us_pc/level_rules.mk levels/level_rules.mk
gmake: cpp-9: No such file or directory
Makefile.split:152: build/us_pc/level_rules.mk: No such file or directory
gmake: *** [Makefile.split:151: build/us_pc/level_rules.mk] Error 127
```

If I go into that levels folder, there are just a bunch of other folders, no 'level_rules.mk' file

**Comment by fgsfdsfgs** [comment]
You're trying to build with clang. It aliases to `gcc` on your system by default. Building with clang is not really supported and you can get GCC from brew. I don't know the details because I don't have an OSX device to test this on. Maybe someone else could help you out because I'm pretty sure you're not the first one to try.

**Comment by mikeymastro** [comment]
Hmmm...ok thanks for the reply.  This is completely foreign to me as I don't know this stuff AT ALL haha, but i'm winging it.  I did download GCC from brew, so I assume I have to change it somewhere I suppose.

**Comment by mikeymastro** [comment]
Wow this is difficult.  I was able to change the version that my MAC uses, and saw that after I ran a `gcc --version` that it was now using the homebrew version.  I ran the entire compiler over again after cloning a fresh copy and I did get further, but now the error is stopping on the same error that i listed in my second comment above:
```

cpp-9 -P -DVERSION_US -DNON_MATCHING -DAVOID_UB -DOSX_BUILD -I . -o build/us_pc/level_rules.mk levels/level_rules.mk
gmake: cpp-9: No such file or directory
Makefile.split:152: build/us_pc/level_rules.mk: No such file or directory
gmake: *** [Makefile.split:151: build/us_pc/level_rules.mk] Error 127
```

Anyone have any ideas?  Not sure why its referencing a file that doesn't exist, so i'm missing something here.  Back to googling for now.  Any help would be appreciated.  Thanks.

**Comment by fgsfdsfgs** [comment]
Yeah, the part of the makefile that deals with OSX_BUILD is actually outdated. Find `CPP := cpp-9` in there and replace with `CPP := cpp`, maybe it'll help.

**Comment by mikeymastro** [comment]
Ahhh man, thanks for your assistance.  That definitely got me further.  It was going along smoothly and then it terminated at the following error:

```
gcc -c -g -I include -I build/us_pc -I build/us_pc/include -I src -I . -DRAPI_GL=1 -DWAPI_SDL2=1 -DAAPI_SDL2=1 -DCAPI_SDL2=1 -DHAVE_SDL2=1 `sdl2-config --cflags` -DVERSION_US -DNON_MATCHING -DAVOID_UB -DOSX_BUILD -DF3DEX_GBI_2E -fno-strict-aliasing -fwrapv -DEXT_OPTIONS_MENU -o build/us_pc/src/pc/pc_main.o src/pc/pc_main.c
src/pc/platform.c:85:10: fatal error: SDL2/SDL.h: No such file or directory
   85 | #include <SDL2/SDL.h>
      |          ^~~~~~~~~~~~
compilation terminated.
gmake: *** [Makefile:1034: build/us_pc/src/pc/platform.o] Error 1
```
ALMOST THERE!  I'll look around for this specific error now in the meantime as well.  Thanks again

**Comment by fgsfdsfgs** [comment]
You have to install SDL2 and maybe other dependencies from brew.

**Comment by mikeymastro** [comment]
Hmmm, I thought I did that with the initial install of brew.  I'll double check that

**Comment by mikeymastro** [comment]
Yea this was my initial install:

`brew install make mingw-w64 gcc sdl2 pkg-config glew glfw3 libusb audiofile git python3`

Is there something else I could run to verify or any other commands that you suggest?

**Comment by mikeymastro** [comment]
I tried to run the install for sdl2 again and got the following:

```
Warning: sdl2 2.0.14_1 is already installed and up-to-date.
To reinstall 2.0.14_1, run:
  brew reinstall sdl2
```

Not really sure what else to do based on my last error.  Googling didn't return much either unfortunately.

**Comment by mikeymastro** [comment]
So since I have no idea what i'm doing, I am thinking I have to somehow point the below code to somehow point to where that SDL.h file is located.  That file is located in /usr/local/Cellar/sdl2/2.0.14_1/include/SDL2/

`gcc -c -g -I include -I build/us_pc -I build/us_pc/include -I src -I . -DRAPI_GL=1 -DWAPI_SDL2=1 -DAAPI_SDL2=1 -DCAPI_SDL2=1 -DHAVE_SDL2=1 `sdl2-config --cflags` -DVERSION_US -DNON_MATCHING -DAVOID_UB -DOSX_BUILD -DF3DEX_GBI_2E -fno-strict-aliasing -fwrapv -DEXT_OPTIONS_MENU -o build/us_pc/src/pc/platform.o src/pc/platform.c`

The above build outputs an error that states the following (just so anyone reading this doesn't have to scroll up):

```
src/pc/platform.c:85:10: fatal error: SDL2/SDL.h: No such file or directory
   85 | #include <SDL2/SDL.h>
      |          ^~~~~~~~~~~~
```

My thing is, where is it looking for this SDL2/SDL.h file?  Line 85?  I can't read the gcc line properly so I have no clue where this is pulling from.

**Comment by mikeymastro** [comment]
Ok so now i'm one step closer.  I see that under src/pc there are a bunch of c files that reference the <SDL2/SDL.h> file in them.  If I remove the SDL2 part and just leave SDL.h in that line of code, it works.  The problem is, it breaks for each c file that is calling that out.  

I'm sure you experts out there know how to change this for all of the files rather than me going 1 by 1, hopefully :)

**Comment by Meowmers101** [comment]
hey, did you ever figure this out? i'm having this problem too.

**Comment by superbonaci** [comment]
I'm having the same issue in OS X Big Sur 11.5.2, on Intel based virtual machine. It's Intel 64 bits both host and guest.

Install HomeBrew and bottles (git is included for some reason):
```
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
brew install make mingw-w64 gcc sdl2 pkg-config glew glfw3 libusb
```

Clean install of sm64ex:
```
git clone https://github.com/sm64pc/sm64ex.git
cd sm64ex
```
Copy `baserom.us.z64` (8388608 bytes):
```
md5sum baserom.us.z64 
20b854b239203baf6c961b850a4a51a2  baserom.us.z64
sha1sum baserom.us.z64 
9bef1128717f958171a4afac3ed78ee2bb4e86ce  baserom.us.z64
```
Now compile:
```
gmake OSX_BUILD=1                                                                              

ld: warning: directory not found for option '-Llib'
ld: warning: option -s is obsolete and being ignored
ld: warning: directory not found for option '-Llib'
ld: warning: option -s is obsolete and being ignored
ld: warning: directory not found for option '-Llib'
ld: warning: option -s is obsolete and being ignored
ld: warning: directory not found for option '-Llib'
ld: warning: option -s is obsolete and being ignored
extracting actors/amp/amp_body.rgba16.png
[...]
extracting textures/water/jrb_textures.0B800.rgba16.png
g++ -std=c++11 -DNDEBUG -fno-rtti -fno-exceptions -fvisibility-inlines-hidden -O2 -I. -c audiofile.cpp
ar rcs libaudiofile.a audiofile.o
gcc -Llib -Iinclude -I../include -I . -Wall -Wextra -Wno-unused-parameter -DOSX_BUILD -pedantic -std=c99 -O2 -s -DOSX_BUILD n64graphics_ci_dir/n64graphics_ci.c n64graphics_ci_dir/exoquant/exoquant.c n64graphics_ci_dir/utils.c -lm -o n64graphics_ci -O2
In file included from n64graphics_ci_dir/n64graphics_ci.c:13:
n64graphics_ci_dir/exoquant/exoquant.h:150:23: warning: no newline at end of file [-Wnewline-eof]
#endif // __EXOQUANT_H
                      ^
In file included from n64graphics_ci_dir/n64graphics_ci.c:15:
n64graphics_ci_dir/n64graphics_ci.h:42:7: warning: no newline at end of file [-Wnewline-eof]
#endif
      ^
2 warnings generated.
In file included from n64graphics_ci_dir/exoquant/exoquant.c:25:
n64graphics_ci_dir/exoquant/exoquant.h:150:23: warning: no newline at end of file [-Wnewline-eof]
#endif // __EXOQUANT_H
                      ^
1 warning generated.
ld: warning: directory not found for option '-Llib'
ld: warning: option -s is obsolete and being ignored
gcc -Llib -Iinclude -I../include -I . -Wall -Wextra -Wno-unused-parameter -DOSX_BUILD -pedantic -std=c99 -O2 -s -DOSX_BUILD n64cksum.c libmio0.c libsm64.c utils.c -lm -o n64cksum 
ld: warning: directory not found for option '-Llib'
ld: warning: option -s is obsolete and being ignored
gcc -Llib -Iinclude -I../include -I . -Wall -Wextra -Wno-unused-parameter -DOSX_BUILD -pedantic -std=c99 -O2 -s -DOSX_BUILD textconv.c utf8.c hashtable.c -lm -o textconv 
ld: warning: directory not found for option '-Llib'
ld: warning: option -s is obsolete and being ignored
gcc -Llib -Iinclude -I../include -I . -Wall -Wextra -Wno-unused-parameter -DOSX_BUILD -pedantic -std=c99 -O2 -s -DOSX_BUILD patch_libultra_math.c -lm -o patch_libultra_math 
ld: warning: directory not found for option '-Llib'
ld: warning: option -s is obsolete and being ignored
gcc -Llib -Iinclude -I../include -I . -Wall -Wextra -Wno-unused-parameter -DOSX_BUILD -pedantic -std=c99 -O2 -s -DOSX_BUILD iplfontutil.c -lm -o iplfontutil -O2
ld: warning: directory not found for option '-Llib'
ld: warning: option -s is obsolete and being ignored
gcc -Llib -Iinclude -I../include -I . -Wall -Wextra -Wno-unused-parameter -DOSX_BUILD -pedantic -std=c99 -O2 -s -DOSX_BUILD aiff_extract_codebook.c -lm -o aiff_extract_codebook 
aiff_extract_codebook.c:48:16: error: implicitly declaring library function 'vsnprintf' with type 'int (char *, unsigned long, const char *, struct __va_list_tag *)' [-Werror,-Wimplicit-function-declaration]
    int size = vsnprintf(NULL, 0, fmt, ap);
               ^
aiff_extract_codebook.c:48:16: note: include the header <stdio.h> or explicitly provide a declaration for 'vsnprintf'
1 error generated.
make: *** [aiff_extract_codebook] Error 1
Makefile:254: *** Failed to build tools.  Stop.
```

@fgsfdsfgs you can create a OS X virtual device with https://github.com/kholia/OSX-KVM which works fantastic under Linux.
Then start with `./OpenCore-Boot.sh`, format a new drive and perform a full install of OS X, then boot into it and do like a regular computer. With `Ctrl+Alt+F` enter and exit fullscreen and with `Ctrl+Alt+G` grab or release focus on virtual machine.

**Comment by Xenthio** [comment]
fixed in #489 (both the sdl2 and the vsnprintf issue)
pls close :)
